hscontrol/api_common.go

```diff
@@ -111,3 +108,71 @@ func (h *Headscale) generateMapResponse(
 
 	return &resp, nil
 }
+
+// mapResponseStreamState tracks state associated with a stream of MapResponse messages,
+// which may optionally send only deltas from the previous message.
+type mapResponseStreamState struct {
+	// peersByID is the peers sent in the last stream message,
+	// for comparison in generating deltas in the new message.
+	peersByID map[uint64]Machine
+}
+
+// applyMapResponseDelta returns a modified MapResponse
+// with fields modified which make use of delta (send on changes).
+//
+// mapResponse the current mapResponse with delta fields not set.
+// streamState optional previous state of mapResponse sent in this stream. Set to nil for a "full update" (no deltas).
+// currentPeers list of peers currently available for the node that this mapResponse is for.
+// toNodes a function to convert the Headscale Machines structure to Tailscale Nodes structure.
+func applyMapResponseDelta(
+	mapResponse tailcfg.MapResponse,
+	streamState *mapResponseStreamState,
+	currentPeers Machines,
+	toNodes func(Machines) ([]*tailcfg.Node, error)) (tailcfg.MapResponse, error,
+) {
+	// Full update
+	if streamState == nil {
```

Comment:
It seems streamState won't be nil when the applyMapResponseDelta called first time in the "case <-updateChan" branch. But there's no peers has been stored in streamState.peersByID. So that may cause the 2nd map response send as delta but generated from an empty state. May that cause some duplicated peers and some peers won't be remove forever in this poll stream?

Reply:
Ah, I think you're right! Good find. This could explain the integration test failure